go/mobile/identity.go

```diff
@@ -0,0 +1,208 @@
+package iden3mobile
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	ethkeystore "github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/iden3/go-iden3-core/components/idenpuboffchain/readerhttp"
+	"github.com/iden3/go-iden3-core/components/idenpubonchain"
+	"github.com/iden3/go-iden3-core/db"
+	"github.com/iden3/go-iden3-core/eth"
+	"github.com/iden3/go-iden3-core/identity/holder"
+	babykeystore "github.com/iden3/go-iden3-core/keystore"
+	"github.com/iden3/go-iden3-crypto/babyjub"
+	log "github.com/sirupsen/logrus"
+)
+
+type Identity struct {
+	id          *holder.Holder
+	Tickets     *TicketsMap
+	eventSender Event
+	storage     db.Storage
+}
+
+// TODO:
+// store on changes vs store on stop
+
+// NewIdentity creates a new identity
+// this funciton is mapped as a constructor in Java.
+// NOTE: The storePath must be unique per Identity
+func NewIdentity(storePath, pass, web3Url string, checkTicketsPeriodMilis int, extraGenesisClaims *BytesArray, e Event) (*Identity, error) {
+	id := &Identity{}
+	idenPubOnChain, keyStore, storage, err := loadComponents(storePath, web3Url)
+	if err != nil {
+		return id, err
+	}
+	resourcesAreClosed := false
+	defer func() {
+		if !resourcesAreClosed {
+			keyStore.Close()
+			storage.Close()
+		}
+	}()
+	// Create babyjub keys
+	kOpComp, err := keyStore.NewKey([]byte(pass))
+	if err != nil {
+		return id, err
+	}
+	if err = keyStore.UnlockKey(kOpComp, []byte(pass)); err != nil {
+		return id, err
+	}
+	// Store kOpComp
+	tx, err := storage.NewTx()
+	if err != nil {
+		return id, err
+	}
+	if err := db.StoreJSON(tx, []byte("kOpComp"), kOpComp); err != nil {
+		return id, err
+	}
+	// Init claim DB
+	tx.Put([]byte("receivedCredentialsCounter"), []byte("0"))
+	if err := tx.Commit(); err != nil {
+		return id, err
+	}
+	// Parse extra genesis claims
+	// TODO: Call toClaimers once it's implemented
+	// _extraGenesisClaims, err := extraGenesisClaims.toClaimers()
+	if err != nil {
+		return id, err
+	}
+	// Create new Identity (holder)
+	_, err = holder.New(
+		holder.ConfigDefault,
+		kOpComp,
+		nil,
+		storage,
+		keyStore,
+		idenPubOnChain,
+		nil,
+		readerhttp.NewIdenPubOffChainHttp(),
+	)
+	if err != nil {
+		return id, err
+	}
+	// Verify that the Identity can be loaded successfully
+	keyStore.Close()
+	storage.Close()
+	resourcesAreClosed = true
+	return NewIdentityLoad(storePath, pass, web3Url, checkTicketsPeriodMilis, e)
+}
+
+// NewIdentityLoad loads an already created identity
+// this funciton is mapped as a constructor in Java
+func NewIdentityLoad(storePath, pass, web3Url string, checkTicketsPeriodMilis int, e Event) (*Identity, error) {
+	// TODO: figure out how to diferentiate the two constructors from Java: https://github.com/iden3/iden3-mobile/issues/17#issuecomment-587374644
+	id := &Identity{}
```

Comment:
Proposal, instead of creating an empty identity here, on line 119 do:
```go
	// Init Identity
	iden := Identity{
		id: holdr,
		Tickets: &TicketsMap{
			m: make(map[string]*Ticket),
		},
		eventSender: e,
		storage: storage,
        }
	iden.loadTickets()
	go iden.checkPendingTickets(time.Duration(checkTicketsPeriodMilis * 1000000))
	return &iden, nil
```

Reply:
cool